llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -16992,6 +17021,33 @@ bool SITargetLowering::checkForPhysRegDependency(
   return false;
 }
 
+/// Check if it is profitable to hoist instruction in then/else to if.
+bool SITargetLowering::isProfitableToHoist(Instruction *I) const {
+  if (!I->hasOneUse())
+    return true;
+
+  Instruction *User = I->user_back();
+  // TODO: Add more patterns that are not profitable to hoist
+  switch (I->getOpcode()) {
+  case Instruction::FMul: {
+    if (User->getOpcode() != Instruction::FSub &&
+        User->getOpcode() != Instruction::FAdd)
+      return true;
+
+    const TargetOptions &Options = getTargetMachine().Options;
+    const Function *F = I->getFunction();
+
+    return ((!I->hasAllowContract() || !User->hasAllowContract()) &&
+            Options.AllowFPOpFusion != FPOpFusion::Fast &&
+            !Options.UnsafeFPMath) ||
```

Comment:
The unsafe fusion mode is an alternative to the fast math flags. They are not both necessary, this should be something like  MissingFlags || FusionMode != Fast

Reply:
I believe existing code takes care of this. This is without pushing the demorgan through -

```c++
return !((I->hasAllowContract() && User->hasAllowContract()) ||
         Options.AllowFPOpFusion == FPOpFusion::Fast ||
         Options.UnsafeFPMath) &&
         isFMAFasterThanFMulAndFAdd(*F, User->getType());
```